### PR TITLE
[Mellanox] Implementation for component version listing

### PR DESCRIPTION
**Pull request by skr31**
#### Why I did it
Added ability to show summary of all the component versions on the switch and a comparison between compiled versions and actual versions. 
The user will be able to run the 'get_component_versions.py' script which outputs the versions table.
There versions table will also be saved in techsupport.

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
* Added Make files that add a new file to the system that contains all the versions of the components available at compilation (the versions are taken from corresponding .mk files).
* Added a script that outputs a table that summarizes all the component versions (compilation and actual)
#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

**Comment by dprital** [comment]
Hi @skr31 , 

Please adjust the subject of this PR as follows:

[202405] [nvidia component listing] Implementation for component version listing -->  [Mellanox] Implementation for component version listing

In addition, please elaborate about the functionality. Meaning:

* How the user will get the information
* Mention that the information will be taken also as part of show techsupport operation

File: files/build_templates/sonic_debian_extension.j2 [hunk 1]
```diff
@@ -973,10 +973,13 @@ for fw_file_name in ${!FW_FILE_MAP[@]}; do
 done
 sudo cp $files_path/$ISSU_VERSION_FILE $FILESYSTEM_ROOT/$PLATFORM_DIR/fw/asic/issu-version
 sudo ln -s /host/image-$SONIC_IMAGE_VERSION/$PLATFORM_DIR/fw/asic/issu-version $FILESYSTEM_ROOT/etc/mlnx/issu-version
+sudo cp $files_path/$COMPONENT_VERSIONS_FILE $FILESYSTEM_ROOT/etc/mlnx/component-versions
 sudo cp $files_path/$MLNX_FFB_SCRIPT $FILESYSTEM_ROOT/usr/bin/mlnx-ffb.sh
 sudo cp $files_path/$MLNX_ONIE_FW_UPDATE $FILESYSTEM_ROOT/usr/bin/$MLNX_ONIE_FW_UPDATE
 sudo cp $files_path/$MLNX_SSD_FW_UPDATE $FILESYSTEM_ROOT/usr/bin/$MLNX_SSD_FW_UPDATE
 sudo cp $files_path/$MLNX_INSTALL_PENDING_FW $FILESYSTEM_ROOT/usr/bin/$MLNX_INSTALL_PENDING_FW
+sudo cp $files_path/$MLNX_GET_COMPONENT_VERSIONS $FILESYSTEM_ROOT/usr/bin/$MLNX_GET_COMPONENT_VERSIONS
```
Comment (dprital) [inline files/build_templates/sonic_debian_extension.j2]: Will it impact other (non-Mellanox) systems ? I am asking because I don't see this area of code is dedicate only for "Mellanox"

File: platform/mellanox/get_component_versions/test/test_get_component_versions.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        self.assertEqual(get_platform_component_versions(), expected_result)
+
+    @patch('os.path.exists')
+    def test_missing_compiled_versions(self, mock_os_path):
+        mock_os_path.return_value = False
+
+        expected_dict = {
+            "SDK": "N/A",
+            "FW": "N/A",
+            "SAI": "N/A",
```
Comment (dprital) [inline platform/mellanox/get_component_versions/test/test_get_component_versions.py]: Why SAI N/A ?

File: files/build_templates/sonic_debian_extension.j2 (hunk 1 above)
Comment (skr31) [reply files/build_templates/sonic_debian_extension.j2]: It is only in Mellanox, there's a {% if sonic_asic_platform == "mellanox" %} above this section

File: platform/mellanox/get_component_versions/test/test_get_component_versions.py (hunk 2 above)
Comment (skr31) [reply platform/mellanox/get_component_versions/test/test_get_component_versions.py]: The test is for missing compiled file, so none of the compiled versions are available, including SAI

**Review by dprital — APPROVED** [review]

File: platform/mellanox/get-component-versions.dep [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# DPKG FRK
+
+DPATH := $($(MLNX_GET_COMPONENT_VERSIONS)_PATH)
+DEP_FILES := $(SONIC_COMMON_FILES_LIST) $(PLATFORM_PATH)/get-component-versions.mk $(PLATFORM_PATH)/get-component-versions.dep
+DEP_FILES += $(SONIC_COMMON_BASE_FILES_LIST)
+DEP_FILES += $(addprefix $(DPATH),$(MLNX_GET_COMPONENT_VERSIONS))
+
+$(MLNX_GET_COMPONENT_VERSIONS)_CACHE_MODE := GIT_CONTENT_SHA
+$(MLNX_GET_COMPONENT_VERSIONS)_DEP_FLAGS := $(SONIC_COMMON_FLAGS_LIST)
+$(MLNX_GET_COMPONENT_VERSIONS)_DEP_FILES := $(DEP_FILES)
```
Comment (nazariig) [inline platform/mellanox/get-component-versions.dep]: @skr31 please drop a new line

File: platform/mellanox/rules.mk [hunk 4]
```diff
@@ -57,4 +59,4 @@ TARGET_BOOTLOADER = grub
 # location for the platform specific external kernel patches
 override EXTERNAL_KERNEL_PATCH_LOC := $(BUILD_WORKDIR)/$(PLATFORM_PATH)/non-upstream-patches/
 
-export SONIC_BUFFER_MODEL=dynamic
+export SONIC_BUFFER_MODEL=dynamic
```
Comment (nazariig) [inline platform/mellanox/rules.mk]: @skr31 please drop a new line

File: platform/mellanox/fw.mk [hunk 5]
```diff
@@ -62,7 +62,8 @@ endif
 
 MLNX_FILES += $(MLNX_FW_FILES)
 
+export MLNX_SPC_FW_VERSION MLNX_SPC2_FW_VERSION MLNX_SPC3_FW_VERSION MLNX_SPC4_FW_VERSION
 export MLNX_SPC_FW_FILE
 export MLNX_SPC2_FW_FILE
 export MLNX_SPC3_FW_FILE
-export MLNX_SPC4_FW_FILE
+export MLNX_SPC4_FW_FILE
```
Comment (nazariig) [inline platform/mellanox/fw.mk]: @skr31 please drop a new line

File: platform/mellanox/component-versions/Makefile [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+MAIN_TARGET = component-versions
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% :
+	echo "SDK $(MLNX_SDK_VERSION)" >> $(DEST)/$(MAIN_TARGET)
+	echo $(MLNX_SPC_FW_VERSION) | sed -r 's/([0-9]*)\.([0-9]*)\.([0-9]*)/FW \2\.\3/g' >> $(DEST)/$(MAIN_TARGET)
+	echo "SAI $(MLNX_SAI_VERSION)" >> $(DEST)/$(MAIN_TARGET)
+	echo "HW-MGMT $(MLNX_HW_MANAGEMENT_VERSION)" >> $(DEST)/$(MAIN_TARGET)
+	echo "MFT $(MFT_VERSION)" >> $(DEST)/$(MAIN_TARGET)
+	echo "Kernel $(KVERSION_SHORT)" >> $(DEST)/$(MAIN_TARGET)
+
```
Comment (nazariig) [inline platform/mellanox/component-versions/Makefile]: @skr31 please drop a new line

File: platform/mellanox/component-versions.mk [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+# Firmware pending update checker and installer
+
+COMPONENT_VERSIONS_FILE = component-versions
+$(COMPONENT_VERSIONS_FILE)_SRC_PATH = $(PLATFORM_PATH)/component-versions
+
+SONIC_MAKE_FILES += $(COMPONENT_VERSIONS_FILE)
+
+MLNX_FILES += $(COMPONENT_VERSIONS_FILE)
+
+export COMPONENT_VERSIONS_FILE
```
Comment (nazariig) [inline platform/mellanox/component-versions.mk]: @skr31 please drop a new line

File: platform/mellanox/fw.mk [hunk 8]
```diff
@@ -62,7 +62,8 @@ endif
 
 MLNX_FILES += $(MLNX_FW_FILES)
 
+export MLNX_SPC_FW_VERSION MLNX_SPC2_FW_VERSION MLNX_SPC3_FW_VERSION MLNX_SPC4_FW_VERSION
 export MLNX_SPC_FW_FILE
 export MLNX_SPC2_FW_FILE
 export MLNX_SPC3_FW_FILE
-export MLNX_SPC4_FW_FILE
```
Comment (nazariig) [inline platform/mellanox/fw.mk]: @skr31 is this a typo? What is the motivation of these changes?

File: platform/mellanox/component-versions/Makefile [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+#
+
+.ONESHELL:
+SHELL = /bin/bash
+.SHELLFLAGS += -e
+
+MAIN_TARGET = component-versions
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% :
+	echo "SDK $(MLNX_SDK_VERSION)" >> $(DEST)/$(MAIN_TARGET)
```
Comment (nazariig) [inline platform/mellanox/component-versions/Makefile]: @skr31 when no cleanup is present, please always have as the first line: `>>` -> `>`

File: platform/mellanox/component-versions/Makefile [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	echo $(MLNX_SPC_FW_VERSION) | sed -r 's/([0-9]*)\.([0-9]*)\.([0-9]*)/FW \2\.\3/g' >> $(DEST)/$(MAIN_TARGET)
+	echo "SAI $(MLNX_SAI_VERSION)" >> $(DEST)/$(MAIN_TARGET)
+	echo "HW-MGMT $(MLNX_HW_MANAGEMENT_VERSION)" >> $(DEST)/$(MAIN_TARGET)
+	echo "MFT $(MFT_VERSION)" >> $(DEST)/$(MAIN_TARGET)
+	echo "Kernel $(KVERSION_SHORT)" >> $(DEST)/$(MAIN_TARGET)
```
Comment (nazariig) [inline platform/mellanox/component-versions/Makefile]: @skr31 please move this logic to a bash script. Once the components file is generated, move it to the relevant destination. Please use `sonic-version.mk` as an example

File: platform/mellanox/get-component-versions.mk [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+#
+# Firmware pending update checker and installer
+
+MLNX_GET_COMPONENT_VERSIONS = get_component_versions.py
+$(MLNX_GET_COMPONENT_VERSIONS)_PATH = $(PLATFORM_PATH)/get_component_versions
+SONIC_COPY_FILES += $(MLNX_GET_COMPONENT_VERSIONS)
+
+MLNX_FILES += $(MLNX_GET_COMPONENT_VERSIONS)
+
+export MLNX_GET_COMPONENT_VERSIONS
```
Comment (nazariig) [inline platform/mellanox/get-component-versions.mk]: @skr31 IMHO, we don't need this extra target just to copy the file. Suggest looking at this example: `https://github.com/sonic-net/sonic-buildimage/pull/14547`. Is it required for running python tests?

File: platform/mellanox/get_component_versions/fwutil/lib.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+class ComponentStatusProvider:
+    def get_status(self):
+        pass
```
Comment (nazariig) [inline platform/mellanox/get_component_versions/fwutil/lib.py]: @skr31 why do we need this? Can we use mock instead?

File: platform/mellanox/get_component_versions/get_component_versions.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    "FW": "N/A",
+    "SAI": "N/A",
+    "HW-MGMT": "N/A",
+    "MFT": "N/A",
+    "Kernel": "N/A"
+}
+
+
+def parse_compiled_components_file():
+    compiled_versions = {}
```
Comment (nazariig) [inline platform/mellanox/get_component_versions/get_component_versions.py]: @skr31 suggest having: `compiled_versions=UNAVAILABLE_COMPILED_VERSIONS`

File: platform/mellanox/get_component_versions/get_component_versions.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+    if not os.path.exists(COMPONENT_VERSIONS_FILE):
+        return UNAVAILABLE_COMPILED_VERSIONS
+
+    with open(COMPONENT_VERSIONS_FILE, 'r') as component_versions:
+        for line in component_versions.readlines():
+            comp, version = line.split()
+            compiled_versions[comp] = version
+
+    if len(compiled_versions.keys()) < len(UNAVAILABLE_COMPILED_VERSIONS.keys()):
```
Comment (nazariig) [inline platform/mellanox/get_component_versions/get_component_versions.py]: @skr31 this doesn't make any sense. Please see the comment above

File: platform/mellanox/get_component_versions/get_component_versions.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+        for component in UNAVAILABLE_COMPILED_VERSIONS.keys():
+            if not compiled_versions.get(component):
+                compiled_versions[component] = "N/A"
+
+    return compiled_versions
+
+
+def get_platform_component_versions():
+    csp = ComponentStatusProvider()
+    version_table = csp.get_status()
```
Comment (nazariig) [inline platform/mellanox/get_component_versions/get_component_versions.py]: @skr31 why do you parse the string output? Suggest using `PlatformDataProvider` instead.
**Note:** Mellanox doesn't have modular chassis.

**Review by nazariig — CHANGES_REQUESTED** [review]

File: platform/mellanox/fw.mk (hunk 8 above)
Comment (skr31) [reply platform/mellanox/fw.mk]: I exported the FW versions so I'll be able to use one of them when creating the component_versions file

File: platform/mellanox/get_component_versions/fwutil/lib.py (hunk 12 above)
Comment (skr31) [reply platform/mellanox/get_component_versions/fwutil/lib.py]: I wasn't able to mock this function without this dummy class

File: platform/mellanox/component-versions/create_component_versions.sh [hunk 16]
```diff
@@ -0,0 +1,6 @@
+echo "SDK $(MLNX_SDK_VERSION)" > $(DEST)/$(MAIN_TARGET)
+echo $(MLNX_SPC_FW_VERSION) | sed -r 's/([0-9]*)\.([0-9]*)\.([0-9]*)/FW \2\.\3/g' >> $(DEST)/$(MAIN_TARGET)
+echo "SAI $(MLNX_SAI_VERSION)" >> $(DEST)/$(MAIN_TARGET)
+echo "HW-MGMT $(MLNX_HW_MANAGEMENT_VERSION)" >> $(DEST)/$(MAIN_TARGET)
+echo "MFT $(MFT_VERSION)" >> $(DEST)/$(MAIN_TARGET)
+echo "Kernel $(KVERSION_SHORT)" >> $(DEST)/$(MAIN_TARGET)
```
Comment (nazariig) [inline platform/mellanox/component-versions/create_component_versions.sh]: @skr31 please drop a new line here

File: platform/mellanox/component-versions/Makefile [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+
+.ONESHELL:
+SHELL = /bin/bash
+.SHELLFLAGS += -e
+
+MAIN_TARGET = component-versions
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% :
+	./create_component_versions.sh
+
```
Comment (nazariig) [inline platform/mellanox/component-versions/Makefile]: @skr31 do we need this extra line?

**Review by nazariig — CHANGES_REQUESTED** [review]

File: platform/mellanox/component-versions/Makefile [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+SHELL = /bin/bash
+.SHELLFLAGS += -e
+
+MAIN_TARGET = component-versions
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% : test-component-versions
+	./create_component_versions.sh > $(DEST)/$(MAIN_TARGET)
+
+test-component-versions:
+	python3 -m unittest discover -s ../get_component_versions/test/
```
Comment (nazariig) [inline platform/mellanox/component-versions/Makefile]: @skr31 why not to use `pytest`? Python UT has a limited functionality

File: platform/mellanox/component-versions/Makefile [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+#
+
+.ONESHELL:
+SHELL = /bin/bash
+.SHELLFLAGS += -e
+
+MAIN_TARGET = component-versions
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% : test-component-versions
+	./create_component_versions.sh > $(DEST)/$(MAIN_TARGET)
```
Comment (nazariig) [inline platform/mellanox/component-versions/Makefile]: @skr31 IMHO, this is a bad design since it is not obvious. Suggest to separate: 1. Make file 2. Move file as separate instruction

File: platform/mellanox/get_component_versions/get_component_versions.py [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+
+    platform_versions = {}
+    for component_name, component in versions_map.items():
+        platform_versions[component_name] = component.get_firmware_version()
+
+    return platform_versions
+
+
+def get_current_version(comp):
+    version = subprocess.run(COMMANDS_FOR_ACTUAL[comp][0], shell=True, executable="/usr/bin/bash", stdout=subprocess.PIPE)
```
Comment (nazariig) [inline platform/mellanox/get_component_versions/get_component_versions.py]: @skr31 what is the use case of `executable="/usr/bin/bash"`? Why do we need it?

File: platform/mellanox/component-versions.mk [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+# Firmware pending update checker and installer
+
+COMPONENT_VERSIONS_FILE = component-versions
+$(COMPONENT_VERSIONS_FILE)_SRC_PATH = $(PLATFORM_PATH)/component-versions
+
+SONIC_MAKE_FILES += $(COMPONENT_VERSIONS_FILE)
+
+MLNX_FILES += $(COMPONENT_VERSIONS_FILE)
+
+$(COMPONENT_VERSIONS_FILE)_DEPENDS += $(SONIC_UTILITIES_PY3)
```
Comment (nazariig) [inline platform/mellanox/component-versions.mk]: @skr31 please put it right after `$(COMPONENT_VERSIONS_FILE)_SRC_PATH`

File: platform/mellanox/get_component_versions/get_component_versions.py [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from fwutil.lib import PlatformDataProvider
+from tabulate import tabulate
+import os
```
Comment (nazariig) [inline platform/mellanox/get_component_versions/get_component_versions.py:19]: @skr31 python imports should be grouped into three groups:
* Standard library imports
* Third-party imports
* Application-specific imports

**Example:**
```python
import a_standard
import b_standard

import a_third_party
import b_third_party

from a_soc import f
from a_soc import g
from b_soc import d
```

**Suggest having:**
```python
import os
import re
import subprocess

from tabulate import tabulate
from fwutil.lib import PlatformDataProvider
```

File: platform/mellanox/get_component_versions/get_component_versions.py [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+        for line in component_versions.readlines():
+            comp, version = line.split()
+            compiled_versions[comp] = version
+
+    return compiled_versions
+
+
+def get_platform_component_versions():
+    pdp = PlatformDataProvider()
+    versions_map = pdp.chassis_component_map.get(next(iter(pdp.chassis_component_map.keys())))
```
Comment (nazariig) [inline platform/mellanox/get_component_versions/get_component_versions.py]: @skr31 it is too complicated. Suggest having something like this:
```python
pdp = PlatformDataProvider()
ccm = pdp.chassis_component_map

if not ccm:
    return UNAVAILABLE_PLATFORM_VERSIONS

for key, value in ccm.items():
    # do stuff here
```

**Review by nazariig — CHANGES_REQUESTED** [review]

File: platform/mellanox/fw.mk (hunk 8 above)
Comment (nazariig) [reply platform/mellanox/fw.mk]: @skr31 ok

**Review by nazariig — APPROVED** [review]
